packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/totp_setup_state_machine.dart

```diff
@@ -59,29 +62,59 @@ final class TotpSetupStateMachine
         )
         .result;
     _session = response.session;
+    _details = TotpSetupDetails(
+      username: CognitoIdToken(tokens.idToken).username,
+      sharedSecret: response.secretCode!,
+    );
     emit(
       TotpSetupState.requiresVerification(
-        TotpSetupDetails(
-          username: CognitoIdToken(tokens.idToken).username,
-          sharedSecret: response.secretCode!,
-        ),
+        _details!,
       ),
     );
   }
 
   Future<void> _onVerify(TotpSetupVerify event) async {
     final tokens = await manager.getUserPoolTokens();
     final accessToken = tokens.accessToken.raw;
-    await _cognitoIdp
-        .verifySoftwareToken(
-          VerifySoftwareTokenRequest(
-            accessToken: accessToken,
-            session: _session,
-            userCode: event.code,
-            friendlyDeviceName: event.friendlyDeviceName,
+    try {
+      await _cognitoIdp
+          .verifySoftwareToken(
+            VerifySoftwareTokenRequest(
+              accessToken: accessToken,
+              session: _session,
+              userCode: event.code,
+              friendlyDeviceName: event.friendlyDeviceName,
+            ),
+          )
+          .result;
+    } on Exception catch (e, st) {
+      // Handle mismatch code exception that may occur during TOTP verification.
+      // See: https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_VerifySoftwareToken.html#API_VerifySoftwareToken_Errors
+      // TODO(equartey): Investigate when CodeMismatchException is thrown and handle accordingly.
+      if (e is EnableSoftwareTokenMfaException) {
+        assert(
+          _details != null,
+          'TotpSetupDetails should not be null. Please report this issue.',
+        );
+        logger.verbose(
+          'Failed to verify TOTP code. Retrying...',
+          e,
+        );
+        emit(
+          TotpSetupState.requiresVerification(
+            _details!,
           ),
-        )
-        .result;
+        );
+        return;
+      }
+      logger.error(
+        'Failed to verify TOTP code. Please try again. If the problem persists, contact support.',
```

Comment:
not sure if we should include `If the problem persists, contact support.` in the error message